arrows/vxl/convert_image.cxx

```diff
@@ -0,0 +1,499 @@
+                        double const random_factor )
+{
+  if( static_cast< double >( rand() ) / RAND_MAX < random_factor )
+  {
+    vil_image_view< Type > compressed;
+    combine_channels( src, compressed );
+
+    vil_image_view< Type > dst{ src.ni(), src.nj(), src.nplanes() };
+
+    for( unsigned p = 0; p < src.nplanes(); ++p )
+    {
+      vil_image_view< Type > output_plane = vil_plane( dst, p );
+      vil_copy_reformat( compressed, output_plane );
+    }
+    return dst;
+  }
+  else
+  {
+    return src;
+  }
+}
+
+// ----------------------------------------------------------------------------
+// Calculate the values of our image percentiles from x sampling points
+template < typename PixType >
+std::vector< PixType >
+sample_and_sort_image( vil_image_view< PixType > const& src,
+                       unsigned int sampling_points,
+                       bool remove_extremes )
+{
+  if( src.ni() * src.nj() < sampling_points )
+  {
+    sampling_points = src.ni() * src.nj();
+  }
+
+  std::vector< PixType > dst;
+
+  if( sampling_points == 0 )
+  {
+    return dst;
+  }
+
+  auto const scanning_area = src.size();
+  auto const ni = src.ni();
+  auto const nj = src.nj();
+  auto const np = src.nplanes();
+  auto const pixel_step = scanning_area / sampling_points;
+
+  dst.resize( sampling_points * np );
+
+  unsigned position = 0;
+
+  for( unsigned p = 0; p < np; ++p )
+  {
+    for( unsigned s = 0;
+         s < sampling_points; ++s, position += pixel_step )
+    {
+      unsigned i = position % ni;
+      unsigned j = ( position / ni ) % nj;
+      dst[ np * s + p ] = src( i, j, p );
+    }
+  }
+
+  std::sort( dst.begin(), dst.end() );
+
+  if( remove_extremes &&
+      dst[ dst.size() - 1 ] == std::numeric_limits< PixType >::max() )
+  {
+    for( unsigned i = 1; i < dst.size(); ++i )
+    {
+      if( dst[ dst.size() - i ] != std::numeric_limits< PixType >::max() )
+      {
+        dst.erase( dst.end() - i, dst.end() );
+        break;
+      }
+    }
+  }
+
+  if( remove_extremes &&
+      dst[ 0 ] == static_cast< PixType >( 0 ) )
+  {
+    for( unsigned i = 1; i < dst.size(); ++i )
+    {
+      if( dst[ i ] != 0 )
+      {
+        dst.erase( dst.begin(), dst.begin() + i );
+        break;
+      }
+    }
+  }
+  return dst;
+}
+
+// ----------------------------------------------------------------------------
+// Estimate the pixel values at given percentiles using a subset of points
+template < typename PixType >
+std::vector< PixType >
+get_image_percentiles(
+  vil_image_view< PixType > const& src,
+  std::vector< double > const& percentiles,
+  unsigned sampling_points, bool remove_extremes )
+{
+  std::vector< PixType > sorted_samples =
+    sample_and_sort_image( src, sampling_points, remove_extremes );
+
+  std::vector< PixType > dst( percentiles.size() );
+  double sampling_points_minus1 =
+    static_cast< double >( sorted_samples.size() - 1 );
+
+  for( unsigned i = 0; i < percentiles.size(); ++i )
```

Comment:
All the `unsigned` here should probably be `size_t`. (Also, this sort of thing is why I love `kwiver::vital::range::iota`; it makes it much easier to use `auto`. :slightly_smiling_face:)